File: LeetCode/remote/longest_palindromic_substring.cs
//  https://leetcode.com/problems/longest-palindromic-substring/
//
//  Given a string S, find the longest palindromic substring in S.
//  You may assume that the maximum length of S is 1000, and there exists one unique longest palindromic substring.

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

public class Solution {
    //  
    //  Submission Details
    //  88 / 88 test cases passed.
    //      Status: Accepted
    //      Runtime: 128 ms
    //          
    //          Submitted: 0 minutes ago
    //
    //          Your runtime beats 95.71% of csharp submissions.
    //          Wow! Holy shit.
    public string LongestPalindrome(string s) {
        var start = -1;
        var length = 0;
        for (var i = 0; i < s.Length; i++)
        {
            var odd = CheckPalindrome(s, i, i);
            var even = CheckPalindrome(s, i, i + 1);
            if (odd * 2 - 1 > length)
            {
                length = odd * 2 - 1;
                start = i - odd + 1;
            }
            if (even * 2 > length)
            {
                length = even * 2;
                start = i - even + 1;
            }
        }

        return s.Substring(start, length);
    }

    //  You are here!
    //  Your runtime beats 1.43% of csharp submissions.
    //
    //  HAHAHA - Parallel
    public string LongestPalindrome2(string s) {
        var start = -1;
        var length = 0;
        Parallel.ForEach(Enumerable.Range(0, s.Length), i => 
        {
            var odd = CheckPalindrome(s, i, i);
            var even = CheckPalindrome(s, i, i + 1);
            if (odd * 2 - 1 > length)
            {
                length = odd * 2 - 1;
                start = i - odd + 1;
            }
            if (even * 2 > length)
            {
                length = even * 2;
                start = i - even + 1;
            }
        });

        return s.Substring(start, length);
    }

    public int CheckPalindrome(String s, int left, int right)
    {
        var length = 0;
        while (left >= 0 && right < s.Length && s[left] == s[right])
        {
            left--;
            right++;
            length++;
        }

        return length;
    }

    static void Main()
    {
        Console.WriteLine(new Solution().LongestPalindrome("abcdcba234"));
        Console.WriteLine(new Solution().LongestPalindrome("abccba2"));
        Console.WriteLine(new Solution().LongestPalindrome2("abcdcba234"));
        Console.WriteLine(new Solution().LongestPalindrome2("abccba2"));
    }
}

File: LeetCode/remote/longest_palindromic_substring.rb
#   https://leetcode.com/problems/longest-palindromic-substring/
#
#   Given a string S, find the longest palindromic substring in S. You may assume that the maximum length of S is 1000,
#   and there exists one unique longest palindromic substring.

#   https://leetcode.com/submissions/detail/54658367/
def longest_palindrome(s)
    max = ""
    0.upto(s.size - 1).each { |i|
        odd = palindrome(s, i, i)
        even = palindrome(s, i, i + 1)
        
        return s if odd.size == s.size || even.size == s.size
        max = odd if odd.size > max.size
        max = even if even.size > max.size
    }

    max    
end

def palindrome(s, low, high)
    while low.between?(0, s.size - 1) &&
          high.between?(0, s.size - 1) &&
          s[low] == s[high]
        low -= 1
        high += 1
    end

    s[low + 1..high - 1]
end

#   DP TLE
def longest_palindrome(s)
    return s if s.to_s.nil?
    palind = Array.new(s.size) {
        Array.new(s.size)
    }  
    
    0.upto(s.size - 1).each { |i|
        palind[i][i] = true
    }
    
    max = s[0]
    0.upto(s.size - 1).each { |i|
        2.upto(s.size - i) { |j|
            if j == i + 1
                palind[i][i + 1] = s[i] == s[i + 1]
            else
                palind[i][j] = s[i] == s[j] && palind[i + 1][j - 1]
            end
            
            max = s[i, j] if palind[i][j] && s[i, j].size > max.size
        }
    }
    
    max
end

